libraries/chain/transaction_context.cpp

```diff
@@ -125,92 +137,104 @@ namespace eosio { namespace chain {
          net_limit_due_to_block = false;
       }
 
-      // Possibly lower objective_duration_limit to optional limit set in transaction header
       if( trx.max_cpu_usage_ms > 0 ) {
          auto trx_specified_cpu_usage_limit = fc::milliseconds(trx.max_cpu_usage_ms);
-         if( trx_specified_cpu_usage_limit <= objective_duration_limit ) {
-            objective_duration_limit = trx_specified_cpu_usage_limit;
-            billing_timer_exception_code = tx_cpu_usage_exceeded::code_value;
-            _deadline = start + objective_duration_limit;
+         if( is_read_only() ) {
+            // Possibly limit deadline to optional limit set in transaction header
+            if( trx_specified_cpu_usage_limit <= read_only_trx_cpu_usage_limit ) {
+               billing_timer_exception_code = tx_cpu_usage_exceeded::code_value;
+               _deadline = start + trx_specified_cpu_usage_limit;
+            }
```

Comment:
If trx specified `max_cpu_usage_ms` is lower than `read_only_trx_cpu_usage_limit` then it would be an objective limit. I think the code should be the same for read_only and normal for `max_cpu_usage_ms`.